File: virtual time capsule/backend/src/main/java/com/java/web/virtual/time/capsule/service/GoalService.java
package com.java.web.virtual.time.capsule.service;

import com.java.web.virtual.time.capsule.dto.GoalCreateDto;
import com.java.web.virtual.time.capsule.dto.GoalDto;
import com.java.web.virtual.time.capsule.dto.UpdateGoalDto;
import com.java.web.virtual.time.capsule.model.Goal;
import java.util.List;

public interface GoalService {
    void createGoal(Long capsuleId, GoalCreateDto goalCreateDto, String creator);
  
    void updateGoal(Long id, UpdateGoalDto updateGoalDto);

    void deleteGoal(Long id);

    List<Goal> getUserGoals(Long userId);

    void setGoalIsAchievedAndMakeVisible(Long id, Boolean achieved);

    GoalDto getGoal(Long id);
}

File: virtual time capsule/frontend/virtual-time-capsule-ui/src/app/services/goal.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { GoalDto } from '../dtos/goal.dto';
import { UpdateGoalDto } from '../dtos/update-goal.dto';
import {environment} from '../../environments/environment';
import {GoalCreateDto} from '../dtos/capsule-create.dto';

@Injectable({
  providedIn: 'root'
})
export class GoalService {
  private apiUrl = `${environment.backendUrl}/api/v1/goals`; // Adjust to your backend URL

  constructor(private http: HttpClient) { }

  getGoal(id: number | null | undefined): Observable<GoalDto> {
    return this.http.get<GoalDto>(`${this.apiUrl}/${id}`);
  }

  createGoal(capsuleId: number, goalDto: GoalCreateDto): Observable<void> {
    let params = new HttpParams().set('capsuleId', capsuleId.toString());
    return this.http.post<void>(`${this.apiUrl}/`, goalDto, { params });
  }

  updateGoal(id: number | null, updateGoalDto: UpdateGoalDto): Observable<void> {
    return this.http.put<void>(`${this.apiUrl}/${id}/update`, updateGoalDto);
  }

  setIsAchieved(id:number | null,isAchieved: boolean): Observable<void> {
    console.log("Sending is Achieved request");
    return this.http.put<void>(
      `${this.apiUrl}/${id}/setIsAchieved?isAchieved=${isAchieved}`,
      null // no request body
    );
  }
}
